### PR TITLE
Datepicker issue3879

### PR DESCRIPTION
**Pull request by thalen**
This solves issue #3879, a problem regarding the datepicker.
I've added a method for date calculation. If a date is selected and the user moves to next or previous month, the date selection is kept if the date is visible in the new view. Otherwise no date is selected.
This also works when navigating via year or month mode.
I've added some tests for this new functionality, I've done some manually testing in Chrome, Firefox and IE9.

Regards,
Olle Thalén

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -370,6 +370,7 @@ module.exports = function(grunt) {
         grunt.config.set('karma.options', karmaOptions);
       }
       grunt.task.run(this.args.length ? 'karma:jenkins' : 'karma:continuous');
+      //grunt.task.run('karma:continuous');
```
Comment (wesleycho) [inline Gruntfile.js]: Remove this change

File: src/datepicker/datepicker.js [hunk 2]
```diff
@@ -76,7 +77,12 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
   }
 
   $scope.isActive = function(dateObject) {
-    if (self.compare(dateObject.date, self.activeDate) === 0) {
+    var compareDate = typeof self.otherMonthSelectedDate !== 'undefined' ? self.otherMonthSelectedDate : self.activeDate;
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: Change this check - one should use `angular.isUndefined`, or a more appropriate check as necessary.

File: src/datepicker/datepicker.js [hunk 3]
```diff
@@ -162,14 +170,82 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
...
+    self.ignoreDateCompare = false;
+    self.otherMonthSelectedDate = undefined;
+  };
+
+  this.calculateDateinterval = function () {
+    var year = self.activeDate.getFullYear();
+    var month = self.activeDate.getMonth();
+    if (typeof ngModelCtrl.$modelValue !== 'undefined' && ngModelCtrl.$modelValue !== null) {
+      var prevValue;
+      if (typeof ngModelCtrl.$modelValue === 'string' || ngModelCtrl.$modelValue instanceof String) {
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: Use `angular.isString` instead

File: src/datepicker/datepicker.js [hunk 4]
```diff
@@ -162,14 +170,82 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
...
 
+  this.clearIntervalVars = function () {
+    self.ignoreDateCompare = false;
+    self.otherMonthSelectedDate = undefined;
+  };
+
+  this.calculateDateinterval = function () {
+    var year = self.activeDate.getFullYear();
+    var month = self.activeDate.getMonth();
+    if (typeof ngModelCtrl.$modelValue !== 'undefined' && ngModelCtrl.$modelValue !== null) {
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: See above on defined checks.

File: src/datepicker/datepicker.js [hunk 5]
```diff
@@ -162,14 +170,82 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
...
+          day = 5;
+        } else if (day > 0) {
+          day = day - 1;
+        }
+        start.setFullYear(year, month, -day);
+      }
+
+      start = start.getTime();
+
+      var nextYear = month === 11 ? year+1 : year;
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: Fix spacing around `+`

File: src/datepicker/datepicker.js [hunk 6]
```diff
@@ -162,14 +170,82 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
...
+      var nextMonth = month+1 % 12;
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: Ditto

File: src/datepicker/datepicker.js [hunk 7]
```diff
@@ -162,14 +170,82 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
...
+      //add days to the end date to get the correct end date for the calendar view
+      var numberofDays = (end-start)/oneDay;
+
+      var days2Add = 42 - Math.round(numberofDays);
+      end = new Date(end);
+      end.setFullYear(nextYear, nextMonth, days2Add);
+      var tmpDate = new Date(start);
+
+      end = end.getTime();
+      if (!(start <= prevValue && prevValue <= end)) {
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: One can do `if (start > prevValue || prevValue > end)` instead

**Comment by wesleycho** [comment]
I'm not sure this solves the problem in an ideal way - this behavior should likely not be hidden behind a config flag for example. What are your thoughts on how it should be solved @Foxandxss ?

File: src/datepicker/datepicker.js (hunk 3 above)
Comment (thalen) [inline src/datepicker/datepicker.js]: Hi,

When I've resolved the issues, what format should I use for the commit message?
Should it be: refactor(datepicker): <msg>

Regards,
Olle Thalén

**Comment by wesleycho** [comment]
Going to close this due to the age - the datepicker has changed quite a lot in the past couple of months, and I'm unsure this particular attempt is the right approach.

Feel free to open a new PR based on current `master`.

**Pull request closed without merging**